examples/dark-mode/app/bootstrap.client.tsx

```diff
@@ -0,0 +1,3 @@
+import { handleDarkAndLightModeEls } from "./utils/theme-provider";
+// fixup stuff before hydration
+handleDarkAndLightModeEls();
```

Comment:
When I was working on this locally, I noticed that this file doesn't get loaded in the browser and this function is not actually run. Can you confirm that for me @mattstobbs? @jacob-ebey, my guess is this would be because we treat all modules as pure and side-effect free?

The objective here is we need this function run as quickly as possible. We need it to happen before importing and evaluating the `react-dom` and `remix` modules in `entry.client.tsx`. Is that not possible or did I just have something weird on my machine?

Reply:
yeah, importing handleDarkAndLightModeEls into entry.client and calling it there will work. That is the only place side-effects are allowed.

Reply:
Why would you want to import and evaluate before react-dom and remix? That makes no sense to me in ESM land. If this was CJS module loading, then sure, but not here. You'd actually need to dynamically import things if this is *really* what you are after, but I don't believe it is.

It also just reads nicer if you import that function and call it directly before they hydrate call showing that this is in fact blocking and executing before react can hydrate.

Reply:
This just needs to happen *very* quickly. Maybe in that case we should move it to an inline script tag instead 🤔 Could probably put it in the other one we've already got. What do you think @mattstobbs?

Reply:
I don't think we could put it in the current one because the body won't have loaded yet so the JavaScript wouldn't be able to find any elements. We could potentially put it on a script tag straight after the Outlet though so it's the first thing that runs?

Reply:
Oh yeah, that's why I didn't include it in the other script tag. Yeah, let's put it in a separate script tag. That will be even better because we don't want to wait for the JS to load.

Another thing to maybe add is: https://github.com/kentcdodds/kentcdodds.com/blob/fa62721066f849e086bc00431982e887f1f26922/styles/app.css#L114-L118

It's possibly something we could do like this:

```css
/* default light, but app-preference is "dark" */
html.dark {
   light-mode {
    display: none;
  }
}

/* default light, and no app-preference */
html:not(.dark) {
  dark-mode {
    display: none;
  }
}

@media (prefers-color-scheme: dark) {
  /* prefers dark, but app-preference is "light" */
  html.light {
    dark-mode {
      display: none;
    }
  }

  /* prefers dark, and app-preference is "dark" */
  html.dark,
  /* prefers dark and no app-preference */
  html:not(.light) {
     light-mode {
      display: none;
    }
  }
}
```

Might even think about inlining that in a style tag as well. Could make sense to have a `<ThemeHead />` and `<ThemeBody />` component.

Another idea is to have everything in a single component people put in the head and then just use a JS event to know when the document is ready to remove all the `<(dark|light)-mode>` elements.

Anyway, there are some ideas for you. Let me know if you get tired of this and I can take it across the finish line for you. Thanks for working on this!

Reply:
Sorry for the long delay, things have been really busy for me. I don't know if I'll get time to work on this this week so if you have time and want to get this over the line, feel free to take over! If not, I'll try and find some time next week!